boojum-cuda/src/poseidon.rs

```diff
@@ -417,31 +420,40 @@ pub fn gather_rows(
 
 pub fn gather_merkle_paths(
     indexes: &DeviceSlice<u32>,
-    values: &(impl DeviceMatrixChunkImpl<BaseField> + ?Sized),
-    result: &mut (impl DeviceMatrixChunkMutImpl<BaseField> + ?Sized),
+    values: &DeviceSlice<GoldilocksField>,
+    results: &mut DeviceSlice<GoldilocksField>,
     stream: &CudaStream,
 ) -> CudaResult<()> {
-    assert_eq!(values.cols(), CAPACITY);
-    assert_eq!(result.cols(), CAPACITY);
+    assert_eq!(values.len() % CAPACITY, 0);
+    let values_count = values.len() / CAPACITY;
+    assert!(values_count.is_power_of_two());
```

Comment:
Technically I don't think this is true, if 'values' is just an unpadded concatenation of size-N, N/2, .... layers up to some cap. But power-of-two size as an API requirement (which may require the caller to add padding) certainly makes internals cleaner.
How is `gather_merkle_paths` called in shivini? https://github.com/search?q=repo%3Amatter-labs%2Fera-shivini+gather_merkle_paths&type=code turns up nothing...maybe it's in your ingonyama demo branch.

Reply:
Indeed, it technically would not require padding but it makes the handling nicer and fits well with the recursive nature of a binary tree.
This layout and padding was introduced with the `build_merkle_tree` function, the `results` parameter of that function and the `values` parameter of  `gather_merkle_paths` are the same.
´gather_merkle_paths´ is not called anywhere yet, it will come as part of the "batched queries" functionality.